content/library/advanced-features/app-testing/beyond-the-basics.md

```diff
@@ -0,0 +1,136 @@
+
+For these attributes, the typical pattern is to declare any values before executing the first run of the app. Values can be inspected at any time in a test. There are a few extra considerations for secrets and Session State which we'll cover now.
+
+## Using secrets with app testing
+
+Be careful not to include secrets directly in your tests. Consider this simple project with `pytest` executed in the project's root directory:
+
+```none
+myproject/
+├── .streamlit/
+│   ├── config.toml
+│   └── secrets.toml
+├── app.py
+└── tests/
+    └── test_app.py
+```
+
+```bash
+cd myproject
+pytest tests/
+```
+
+In the above scenario, your simulated app will have access to your `secrets.toml` file. However, since you don't want to commit your secrets to your repository, you may need to write tests where you securely pull your secrets into memory or use dummy secrets.
+
+### Example: declaring secrets in a test
+
+Within a test, declare each secret after initializing your `AppTest` instance but before the first run. (A missing secret may result in an app that doesn't run!) For example, consider the following secrets file and corresponding test initialization to manually assign the same secrets:
+
+Secrets file:
+
+```toml
+db_username = "Jane"
+db_password = "mypassword"
+
+[my_other_secrets]
+things_i_like = ["Streamlit", "Python"]
+```
+
+Testing file with equivalent secrets:
+
+```python
+# Initialize an AppTest instance.
+at = AppTest.from_file("app.py")
+# Declare the secrets.
+at.secrets["db_username"] = "Jane"
+at.secrets["db_password"] = "mypassword"
+at.secrets["my_other_secrets.things_i_like"] = ["Streamlit", "Python"]
+# Run the app.
+at.run()
+```
+
+Generally, you don't want to type your secrets directly into your test. If you don't need your real secrets for a test, you can declare dummy secrets. If your tests require your real secrets, you should use an API to pass them securely and anonymously. If you are automating your tests with GitHub actions, check out their [Security guide](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).
+
+```python
+at.secrete["my_key"] = <value provided through API>
+```
+
+## Working with Session State in app testing
+
+Just like with secrets, the `.session_state` attribute for `AppTest` let's you read and update Session State values using key notation (`at.session_state["my_key"]`), but not attribute notation (`at.session_state.my_key`). By manually declaring values in Session State, you can directly jump to a specific state instead of simulating many steps to get there. Additionally, the testing framework does not provide native support for multipage apps. An instance of `AppTest` can only test one page. You need to manually declare Session State values if you want to simulate a user having data from another page.
```

Comment:
I thought `at.session_state` actually does support attribute notation (unlike the other 2) but I could be wrong.

Also might add something like: "This also allows you to inspect the contents of session_state after a run to confirm an update happened as expected."

Reply:
I just tested, and you are indeed correct. Sorry I missed that one. Fixing it now.